Include/Library/OcCryptoLib.h

```diff
@@ -305,6 +393,118 @@ Sha384 (
   UINTN        Len
   );
 
+/**
+  Verify a RSA PKCS1.5 signature against an expected hash.
+
+  @param[in] Modulus        The RSA modulus byte array.
+  @param[in] ModulusSize    The size, in bytes, of Modulus.
+  @param[in] Exponent       The RSA exponent.
+  @param[in] Signature      The RSA signature to be verified.
+  @param[in] SignatureSize  Size, in bytes, of Signature.
+  @param[in] Hash           The Hash digest of the signed data.
+  @param[in] HashSize       Size, in bytes, of Hash.
+  @param[in] Algorithm      The RSA algorithm used.
+
+  @returns  Whether the signature has been successfully verified as valid.
+
+**/
+BOOLEAN
+RsaVerifyFromKey (
```

Comment:
The naming scheme I took from the old draft, I thought the separation of RSA (general purpose) and Signature Verification (only that obviously) was desired. Though the RSA function is not really generic in the first place, so I guess it's indeed better to adapt the naming.

Reply:
Regarding different cryptography implementations, I'm not convinced it makes much sense to design the interfaces generically when we do not know the requirements of potential other cryptography methods. For now, I think it should be clear RSA is used.

Reply:
Regarding cipher strength, I expected we trust the Apple Root Certificate considering it uses a strong cipher itself. And as for the user, I expected we leave cipher choice to them. However, I also agree we can add a minimum strentgh, such as 2048 Bits maybe?

Reply:
Naming adapted at https://github.com/acidanthera/OcSupportPkg/commit/33b60c588a3880323589754fe18084a4c31bfc90

Reply:
Regarding the adaptation from the industry practice we can estimate that fairly well, I would say ECDSA and ECDH Curve25519 are the most common things used for signature verification along with RSA. For them the interface would remain mostly the same: pubkey, data/hash, signature(s). Sure, we will have to store that pubkey in a different format, so the key argument can become a union in the future or a list of keys (which is somewhat more realistic). For now I am fine to assume that we are sticking with rsa, and not adapt the naming scheme, but I personally see no barrier for that.

Regarding key size I want to see at any place calling cryptography that it uses only trusted combinations. You might not be aware, but basically all software using crypto have a configuration whitelist of primitive combinations that are allowed to be used. See SSLCipherSuite e.g. [here](https://httpd.apache.org/docs/2.4/ssl/ssl_howto.html). I see two ways to do it:
- pass it into the functions (and I wanted to use that Algorithm field for that, but most likely it is an abuse)
- define it during compilation with macros or pcds

Reply:
I was aware of the idea, but not any specifics, thanks.
For our usage cases, I still do not see a reason for ciphers "parameters" being limited upwards. I still think we want a minimum, but of course it does not need to be an arbitrary integer or such (considering the interface). If we had enum values for RSA-2048, RSA-4096 or such, but used the modulus size as a minimum, what would be the issue with that? Afterall it's literally only the modulus size that is changing.
A whitelist would make more sense to me when dealing with something like SHA, where the output size is just one aspect of the algorithm - if one SHA function was vulnerable to an arbitrary attack, it might make sense to not include it individually.

Reply:
I will have to repeat that that just putting some bounds like lower limit does not work in cryptography. You perform validation of certain cypher suits and combinations with the means required from you by the regulator or your own will. Then you declare that these work and pass the certification if required.

Removing the upper limit basically is impossible, as you need to validate and guarantee the infinite amount of options. Putting a very large upper limit also does not work, as you still need to test all possible combinations, and ensure, that they do not have potential vulnerabilities, do not cause denial of service, execute in adequate timespan etc. It also does not work in practice, AES has 256-bit keys because it makes sense, so does RSA with 4096-bit keys. If an algorithm needs a serious key size increase to become secure it is about the time to change the algorithm, see [first answer](https://crypto.stackexchange.com/questions/1182/are-there-practical-upper-limits-of-rsa-key-lengths).

So no, I do Indeed require "artificial" whitelist, that is statically configurable on OcCryptoLib compilation.